tests/federation_room_join_partial_state_test.go

```diff
@@ -121,7 +121,14 @@ func TestPartialStateJoin(t *testing.T) {
 	) {
 		t.Helper()
 
-		user.MustSyncUntil(t, client.SyncReq{}, client.SyncJoinedTo(user.UserID, room.RoomID))
+		// Use a `/members` request to wait for the room to be un-partial stated.
+		// We avoid using `/sync`, as it only waits (or used to wait) for full state at
+		// particular events, rather than the whole room.
+		user.MustDoFunc(
+			t,
+			"GET",
+			[]string{"_matrix", "client", "v3", "rooms", room.RoomID, "members"},
+		)
```

Comment:
I wonder if we should expose a way for clients to know if the room is partial, rather than trying to infer it? It even sounds like something we might want to expose in the UI?

Reply:
That sounds possibly useful. Though I wonder how much the concept of a partial state room (as opposed to a partial state event) is specific to the synapse implementation.

I suppose there are at least three reasons `awaitPartialStateJoinCompletion` might be used:
* to wait for a particular event to have full state
* to wait for the "current" state of the room to be a full state
* to wait for all events in the room to have full state (implies the previous 2)

I'd guess that clients and most tests are primarily interested in the 2nd property and it doesn't matter if there's still a partial state event somewhere back in the dag (which synapse doesn't allow, but a theoretical homeserver implementation might).